File: feedtamer/backend/src/models/content.model.js
const mongoose = require('mongoose');

const contentSchema = new mongoose.Schema({
  account: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Account',
    required: true
  },
  platform: {
    type: String,
    required: true,
    enum: ['twitter', 'instagram', 'youtube', 'reddit', 'linkedin']
  },
  contentId: {
    type: String,
    required: true
  },
  contentType: {
    type: String,
    required: true,
    enum: ['post', 'tweet', 'photo', 'video', 'story', 'article', 'comment']
  },
  text: {
    type: String
  },
  mediaUrls: [{
    type: String
  }],
  mediaType: {
    type: String,
    enum: ['image', 'video', 'gif', 'audio', 'link', 'none']
  },
  link: {
    type: String
  },
  publishedAt: {
    type: Date,
    required: true
  },
  retrievedAt: {
    type: Date,
    default: Date.now
  },
  engagementStats: {
    likes: Number,
    comments: Number,
    shares: Number,
    views: Number
  },
  users: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    seen: {
      type: Boolean,
      default: false
    },
    seenAt: Date,
    saved: {
      type: Boolean,
      default: false
    },
    savedAt: Date,
    interaction: {
      liked: Boolean,
      commented: Boolean,
      shared: Boolean
    }
  }],
  rawData: {
    type: mongoose.Schema.Types.Mixed
  }
});

// Compound index to ensure uniqueness of content per platform
contentSchema.index({ platform: 1, contentId: 1 }, { unique: true });

// Index for efficient feed generation queries
contentSchema.index({ account: 1, publishedAt: -1 });
contentSchema.index({ platform: 1, publishedAt: -1 });

// Index for user-specific content queries
contentSchema.index({ 'users.user': 1, publishedAt: -1 });
contentSchema.index({ 'users.user': 1, 'users.saved': 1 });

// Instance method to add or update user interaction with content
contentSchema.methods.updateUserInteraction = async function(userId, interactionData) {
  const userIndex = this.users.findIndex(u => u.user.toString() === userId.toString());
  
  if (userIndex >= 0) {
    // Update existing record
    if (interactionData.seen) {
      this.users[userIndex].seen = true;
      this.users[userIndex].seenAt = new Date();
    }
    
    if (interactionData.saved !== undefined) {
      this.users[userIndex].saved = interactionData.saved;
      if (interactionData.saved) {
        this.users[userIndex].savedAt = new Date();
      }
    }
    
    if (interactionData.interaction) {
      this.users[userIndex].interaction = {
        ...this.users[userIndex].interaction,
        ...interactionData.interaction
      };
    }
  } else {
    // Add new user record
    this.users.push({
      user: userId,
      seen: interactionData.seen || false,
      seenAt: interactionData.seen ? new Date() : undefined,
      saved: interactionData.saved || false,
      savedAt: interactionData.saved ? new Date() : undefined,
      interaction: interactionData.interaction || {}
    });
  }
  
  return this.save();
};

// Static method to get content for a user's feed
contentSchema.statics.getFeedForUser = async function(userId, accounts, options = {}) {
  const {
    limit = 20,
    skip = 0,
    platform = null,
    onlySaved = false,
    includeRead = false,
    maxAge = null
  } = options;
  
  const query = {
    account: { $in: accounts }
  };
  
  if (platform) {
    query.platform = platform;
  }
  
  if (maxAge) {
    const maxAgeDate = new Date();
    maxAgeDate.setDate(maxAgeDate.getDate() - maxAge);
    query.publishedAt = { $gte: maxAgeDate };
  }
  
  if (onlySaved) {
    query['users.user'] = userId;
    query['users.saved'] = true;
  }
  
  if (!includeRead) {
    // Exclude content this user has already seen
    query.$or = [
      { 'users.user': { $ne: userId } },
      { 'users': { $elemMatch: { user: userId, seen: false } } }
    ];
  }
  
  return this.find(query)
    .sort({ publishedAt: -1 })
    .skip(skip)
    .limit(limit)
    .populate('account', 'username displayName profilePicture platform');
};

const Content = mongoose.model('Content', contentSchema);

module.exports = Content;

File: feedstamer/flutter_app/lib/models/content_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:Feedstamer/models/account_model.dart';

class ContentModel {
  final String id;
  final String accountId;
  final AccountModel? account;
  final String platform;
  final String contentId;
  final String contentType; // post, tweet, photo, video, story, article, comment
  final String? text;
  final List<String> mediaUrls;
  final String? mediaType; // image, video, gif, audio, link, none
  final String? link;
  final DateTime publishedAt;
  final DateTime retrievedAt;
  final ContentEngagementStats? engagementStats;
  final List<ContentUserInteraction> userInteractions;
  final Map<String, dynamic>? rawData;
  
  ContentModel({
    required this.id,
    required this.accountId,
    this.account,
    required this.platform,
    required this.contentId,
    required this.contentType,
    this.text,
    required this.mediaUrls,
    this.mediaType,
    this.link,
    required this.publishedAt,
    required this.retrievedAt,
    this.engagementStats,
    required this.userInteractions,
    this.rawData,
  });
  
  factory ContentModel.fromMap(Map<String, dynamic> map, {required String id, AccountModel? account}) {
    List<ContentUserInteraction> interactions = [];
    
    if (map['userInteractions'] != null) {
      interactions = (map['userInteractions'] as List)
        .map((item) => ContentUserInteraction.fromMap(item))
        .toList();
    }
    
    return ContentModel(
      id: id,
      accountId: map['accountId'] ?? '',
      account: account,
      platform: map['platform'] ?? '',
      contentId: map['contentId'] ?? '',
      contentType: map['contentType'] ?? '',
      text: map['text'],
      mediaUrls: List<String>.from(map['mediaUrls'] ?? []),
      mediaType: map['mediaType'],
      link: map['link'],
      publishedAt: (map['publishedAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      retrievedAt: (map['retrievedAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      engagementStats: map['engagementStats'] != null 
        ? ContentEngagementStats.fromMap(map['engagementStats']) 
        : null,
      userInteractions: interactions,
      rawData: map['rawData'],
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'accountId': accountId,
      'platform': platform,
      'contentId': contentId,
      'contentType': contentType,
      'text': text,
      'mediaUrls': mediaUrls,
      'mediaType': mediaType,
      'link': link,
      'publishedAt': Timestamp.fromDate(publishedAt),
      'retrievedAt': Timestamp.fromDate(retrievedAt),
      'engagementStats': engagementStats?.toMap(),
      'userInteractions': userInteractions.map((item) => item.toMap()).toList(),
      'rawData': rawData,
    };
  }
  
  ContentModel copyWith({
    String? id,
    String? accountId,
    AccountModel? account,
    String? platform,
    String? contentId,
    String? contentType,
    String? text,
    List<String>? mediaUrls,
    String? mediaType,
    String? link,
    DateTime? publishedAt,
    DateTime? retrievedAt,
    ContentEngagementStats? engagementStats,
    List<ContentUserInteraction>? userInteractions,
    Map<String, dynamic>? rawData,
  }) {
    return ContentModel(
      id: id ?? this.id,
      accountId: accountId ?? this.accountId,
      account: account ?? this.account,
      platform: platform ?? this.platform,
      contentId: contentId ?? this.contentId,
      contentType: contentType ?? this.contentType,
      text: text ?? this.text,
      mediaUrls: mediaUrls ?? this.mediaUrls,
      mediaType: mediaType ?? this.mediaType,
      link: link ?? this.link,
      publishedAt: publishedAt ?? this.publishedAt,
      retrievedAt: retrievedAt ?? this.retrievedAt,
      engagementStats: engagementStats ?? this.engagementStats,
      userInteractions: userInteractions ?? this.userInteractions,
      rawData: rawData ?? this.rawData,
    );
  }
  
  // Helper methods
  bool isSeenByUser(String userId) {
    final interaction = userInteractions.firstWhere(
      (item) => item.userId == userId,
      orElse: () => ContentUserInteraction(userId: userId),
    );
    return interaction.seen;
  }
  
  bool isSavedByUser(String userId) {
    final interaction = userInteractions.firstWhere(
      (item) => item.userId == userId,
      orElse: () => ContentUserInteraction(userId: userId),
    );
    return interaction.saved;
  }
  
  // Platform-specific icon
  String get platformIcon {
    switch (platform) {
      case 'twitter':
        return 'assets/icons/twitter.png';
      case 'instagram':
        return 'assets/icons/instagram.png';
      case 'youtube':
        return 'assets/icons/youtube.png';
      case 'reddit':
        return 'assets/icons/reddit.png';
      case 'linkedin':
        return 'assets/icons/linkedin.png';
      default:
        return 'assets/icons/link.png';
    }
  }
  
  // For displaying timeframe
  String getTimeAgo() {
    final now = DateTime.now();
    final difference = now.difference(publishedAt);
    
    if (difference.inSeconds < 60) {
      return 'Just now';
    } else if (difference.inMinutes < 60) {
      return '${difference.inMinutes}m ago';
    } else if (difference.inHours < 24) {
      return '${difference.inHours}h ago';
    } else if (difference.inDays < 7) {
      return '${difference.inDays}d ago';
    } else if (difference.inDays < 30) {
      final weeks = (difference.inDays / 7).floor();
      return '${weeks}w ago';
    } else if (difference.inDays < 365) {
      final months = (difference.inDays / 30).floor();
      return '${months}mo ago';
    } else {
      final years = (difference.inDays / 365).floor();
      return '${years}y ago';
    }
  }
}

class ContentEngagementStats {
  final int? likes;
  final int? comments;
  final int? shares;
  final int? views;
  
  ContentEngagementStats({
    this.likes,
    this.comments,
    this.shares,
    this.views,
  });
  
  factory ContentEngagementStats.fromMap(Map<String, dynamic> map) {
    return ContentEngagementStats(
      likes: map['likes'],
      comments: map['comments'],
      shares: map['shares'],
      views: map['views'],
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'likes': likes,
      'comments': comments,
      'shares': shares,
      'views': views,
    };
  }
}

class ContentUserInteraction {
  final String userId;
  final bool seen;
  final DateTime? seenAt;
  final bool saved;
  final DateTime? savedAt;
  final Map<String, bool>? interaction;
  
  ContentUserInteraction({
    required this.userId,
    this.seen = false,
    this.seenAt,
    this.saved = false,
    this.savedAt,
    this.interaction,
  });
  
  factory ContentUserInteraction.fromMap(Map<String, dynamic> map) {
    return ContentUserInteraction(
      userId: map['userId'] ?? '',
      seen: map['seen'] ?? false,
      seenAt: (map['seenAt'] as Timestamp?)?.toDate(),
      saved: map['saved'] ?? false,
      savedAt: (map['savedAt'] as Timestamp?)?.toDate(),
      interaction: map['interaction'] != null
        ? Map<String, bool>.from(map['interaction'])
        : null,
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'userId': userId,
      'seen': seen,
      'seenAt': seenAt != null ? Timestamp.fromDate(seenAt!) : null,
      'saved': saved,
      'savedAt': savedAt != null ? Timestamp.fromDate(savedAt!) : null,
      'interaction': interaction,
    };
  }
}